tests/utils/atomFamily.test.tsx

```diff
@@ -0,0 +1,116 @@
+import React from 'react'
+import { fireEvent, render } from '@testing-library/react'
+import { Provider, atom, useAtom } from '../../src/index'
+import { atomFamily } from '../../src/utils'
+import { SetStateAction, PrimitiveAtom } from '../../src/core/types'
+
+it('atomFamily returns same reference for same parameters', async () => {
+  const arrayAtom = atom([0])
+  const myFamily = atomFamily<number, number>((num) => (get) =>
+    get(arrayAtom)[num]
+  )
+  expect(myFamily(0)).toEqual(myFamily(0))
+  expect(myFamily(0)).not.toEqual(myFamily(1))
+  expect(myFamily(1)).not.toEqual(myFamily(0))
+})
+
+it('removed atom creates a new reference', async () => {
+  const bigAtom = atom([0])
+  const myFamily = atomFamily<number, number>((num) => (get) =>
+    get(bigAtom)[num]
+  )
+
+  const savedReference = myFamily(0)
+
+  expect(savedReference).toEqual(myFamily(0))
+
+  myFamily.remove(0)
+
+  const newReference = myFamily(0)
+
+  expect(savedReference).not.toEqual(newReference)
+
+  myFamily.remove(1337)
+
+  expect(myFamily(0)).toEqual(newReference)
+})
+
+it('atomFamily functionality as usual', async () => {
+  const arrayAtom = atom([0, 0, 0])
+
+  const myFamily = atomFamily<number, number, SetStateAction<number>>(
+    (param) => (get) => get(arrayAtom)[param],
+    (param) => (_, set, update) => {
+      set(arrayAtom, (oldArray) => {
+        if (typeof oldArray[param] === 'undefined') return oldArray
+
+        const newValue =
+          typeof update === 'function' ? update(oldArray[param]) : update
+
+        const newArray = [
+          ...oldArray.slice(0, param),
+          newValue,
+          ...oldArray.slice(param + 1),
+        ]
+
+        return newArray
+      })
+    }
+  )
+
+  const Displayer = ({
+    index,
+    countAtom,
+  }: {
+    index: number
+    countAtom: PrimitiveAtom<number>
+  }) => {
+    const [count, setCount] = useAtom(countAtom)
+    return (
+      <div>
+        index: {index}, count: {count}
+        <button onClick={() => setCount((oldValue) => oldValue + 1)}>
+          increment #{index}
+        </button>
+      </div>
+    )
+  }
+
+  const Parent: React.FC = () => {
+    const [indices] = useAtom(
+      atom((get) => [...new Array(get(arrayAtom).length)])
```

Comment:
Is it intentional to create an atom in each render, instead of creating it outside component?

Reply:
Well, yeah... I think of it "it's the parents responsibility to create it, no one else is interested in it".

Should have I created it outside?

Is this a possible memory leak?

Reply:
> Should have I created it outside?

Yes, please. It's more idiomatic.

> Is this a possible memory leak?

It doesn't leak memory, but every time the component renders, it will run initialization and garbage collection. So, it's a performance issue.

Reply:
> It doesn't leak memory, but every time the component renders, it will run initialization and garbage collection. So, it's a performance issue.

I understand. Fortunately, in this example there's no way to make Parent re-render ;)

But yes, I think this answers another question I've had about jotai, the way I've been using it there's this "commit" in the performance tab of react tools, this commit always comes before the real component update, that might be the initialization/garbage-collection commit!